src/System.Buffers/src/System/Buffers/BufferData.cs

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace System.Buffers
+{
+    public struct BufferData
+    {
+        internal byte[] _data;
+        internal bool _allocd;
```

Comment:
I'd just drop extra buffers on the floor. 
You are already dropping buffers when pool is oversubscribed. That is unavoidable since you have a fixed size pool but potentially unbounded demand for buffers (however unlikely).
I only suggest to drop under different conditions - not when a buffer was not from the pool originally, but when pool does not have space to accept it back.